rustdoc-search: add support for type parameters

notriddle opened the pull request:
r? @GuillaumeGomez 

## Preview

* https://notriddle.com/rustdoc-html-demo-4/advanced-search/rustdoc/read-documentation/search.html
* https://notriddle.com/rustdoc-html-demo-4/advanced-search/std/index.html?search=option%3Coption%3CT%3E%3E%20-%3E%20option%3CT%3E
* https://notriddle.com/rustdoc-html-demo-4/advanced-search/std/index.html?search=option%3CT%3E,%20E%20-%3E%20result%3CT,%20E%3E
* https://notriddle.com/rustdoc-html-demo-4/advanced-search/std/index.html?search=-%3E%20option%3CT%3E

## Description

When writing a type-driven search query in rustdoc, specifically one with more than one query element, non-existent types become generic parameters instead of auto-correcting (which is currently only done for single-element queries) or giving no result. You can also force a generic type parameter by writing `generic:T` (and can force it to not use a generic type parameter with something like `struct:T` or whatever, though if this happens it means the thing you're looking for doesn't exist and will give you no results).

There is no syntax provided for specifying type constraints for generic type parameters.

When you have a generic type parameter in a search query, it will only match up with generic type parameters in the actual function, not concrete types that match, not concrete types that implement a trait. It also strictly matches based on when they're the same or different, so `option<T>, option<U> -> option<U>` matches `Option::and`, but not `Option::or`. Similarly, `option<T>, option<T> -> option<T>` matches `Option::or`, but not `Option::and`.

## Motivation

This feature is motivated by the many "combinitor"-type functions found in generic libraries, such as Option, Future, Iterator, and Entry. These highly-generic functions have names that are almost completely arbitrary, and a type signature that tells you what it actually does.

This PR is a major step towards[^closure] being able to easily search for generic functions by their type signature instead of by name. Some examples of combinators that can be found using this PR (try them out in the preview):

* `option<option<T>> -> option<T>` returns Option::flatten
* `option<T> -> result<T>` returns Option::ok_or
* `option<result<T>> -> result<option<T>>` returns Option::transpose
* `entry<K, V>, FnOnce -> V` returns `Entry::or_insert_with` (and `or_insert_with_key`, since there's no way to specify the generics on FnOnce)

[^closure]:

    For this feature to be as useful as it ought to be, you should be able to search for *trait-associated types* and *closures*. This PR does not implement either of these: they are **Future possibilities**.

    Trait-associated types would allow queries like `option<T> -> iterator<item=T>` to return `Option::iter`. We should also allow `option<T> -> iterator<T>` to match the associated type version.

    Closures would make a good way to query for things like `Option::map`. Closure support needs associated types to be represented in the search index, since `FnOnce() -> i32` desugars to `FnOnce<Output=i32, ()>`, so associated trait types should be implemented first. Also, we'd want to expose an easy way to query closures without specifying which of the three traits you want.

notriddle committed:
rustdoc-search: `null`, not `-1`, for missing id

This allows us to use negative numbers for others purposes.

notriddle committed:
rustdoc-search: add support for type parameters

When writing a type-driven search query in rustdoc, specifically one
with more than one query element, non-existent types become generic
parameters instead of auto-correcting (which is currently only done
for single-element queries) or giving no result. You can also force a
generic type parameter by writing `generic:T` (and can force it to not
use a generic type parameter with something like `struct:T` or whatever,
though if this happens it means the thing you're looking for doesn't
exist and will give you no results).

There is no syntax provided for specifying type constraints
for generic type parameters.

When you have a generic type parameter in a search query, it will only
match up with generic type parameters in the actual function, not
concrete types that match, not concrete types that implement a trait.
It also strictly matches based on when they're the same or different,
so `option<T>, option<U> -> option<U>` matches `Option::and`, but not
`Option::or`. Similarly, `option<T>, option<T> -> option<T>`` matches
`Option::or`, but not `Option::and`.

notriddle committed:
rustdoc: write detailed chapter on search engine

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5295579944/jobs/9586044244) [(plain)](https://github.com/rust-lang/rust/commit/cc15381125adafce85c7889ca1d99c6a28f21cc5/checks/14337500322/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Rustbook (x86_64-unknown-linux-gnu) - style-guide
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.10s
##[endgroup]
rustdoc/read-documentation/search.html:196: broken intra-doc link - [<code>stdout, [u8]</code>]
rustdoc/read-documentation/search.html:200: broken intra-doc link - [<code>stdout, [u8]</code>]
rustdoc/print.html:673: broken intra-doc link - [<code>stdout, [u8]</code>]
rustdoc/print.html:677: broken intra-doc link - [<code>stdout, [u8]</code>]
number of HTML files scanned: 33986
number of HTML redirects found: 10364
number of links checked: 2968806
number of links ignored due to external: 57377
```

</details>

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -1267,44 +1223,98 @@ function initSearch(rawSearchIndex) {
+             *
+             * @type Map<integer, FunctionType[]>
+             */
+            const unbox = function unbox(fnType) {
```
No need for a named function here.

```suggestion
            const unbox = function (fnType) {
```

GuillaumeGomez commented:
If I remember correctly, having exact match when more than one element is present in the search query is voluntary. Unless I misunderstood what this PR is about?

cc @jsha too

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5312778241/jobs/9617774158) [(plain)](https://github.com/rust-lang/rust/commit/e46723a21e626097026df62540b7f12801ac2bb4/checks/14376097907/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: ES-Check: there were no ES version matching errors!  🎉
+ eslint -c ../src/librustdoc/html/static/.eslintrc.js ../src/librustdoc/html/static/js/externs.js ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/source-script.js ../src/librustdoc/html/static/js/storage.js

/checkout/src/librustdoc/html/static/js/search.js
  1277:35  error  Unexpected space before function parentheses  space-before-function-paren

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

</details>

GuillaumeGomez commented:
Ok, thanks for the clarification, I misunderstood your explanations.

GuillaumeGomez commented on src/doc/rustdoc/src/read-documentation/search.md:
```diff
@@ -0,0 +1,237 @@
+[`vec vec`]: ../../std/vec/struct.Vec.html?search=vec%20vec&filter-crate=std
+[`std::vec`]: ../../std/vec/struct.Vec.html?search=std::vec&filter-crate=std
+[`std::vec::Vec`]: ../../std/vec/struct.Vec.html?search=std::vec::Vec&filter-crate=std
+[`std::vec::Vec`]: ../../std/vec/struct.Vec.html?search=std::vec::Vec&filter-crate=std
```
This one is present twice.

GuillaumeGomez commented on src/doc/rustdoc/src/read-documentation/search.md:
````diff
@@ -0,0 +1,237 @@
+
+## Search query syntax
+
+```text
````
Please add a sentence explaining it's the eBNF for the search syntax with a link to a definition of what "eBNF" means (wikipedia?).

GuillaumeGomez replied:
:+1:

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -387,96 +399,48 @@ fn get_index_type_id(clean_type: &clean::Type) -> Option<RenderTypeId> {
+}
+
+/// The point of this function is to lower generics and types into the simplified form that the
+/// frontend search engine can use.
```
"front-end" ;)

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -387,96 +399,48 @@ fn get_index_type_id(clean_type: &clean::Type) -> Option<RenderTypeId> {
-fn add_generics_and_bounds_as_types<'tcx, 'a>(
+/// This function also works recursively.
+#[instrument(level = "trace", skip(tcx, res, rgen, cache))]
+fn simplify_fn_type<'tcx, 'a>(
```
Considering the number of arguments, wouldn't it be worth creating a struct with these fields so at least we can know from where the function is called which is which?

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -1267,44 +1223,98 @@ function initSearch(rawSearchIndex) {
+                        }
+                    }
+                }
+                if (qsid === null || !(
```
Took me a while to mentally parse this condition. Would it be possible to move it into a function and have conditions split to make reading easier? An inner function is completely fine.

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -1344,15 +1356,32 @@ function initSearch(rawSearchIndex) {
-                    if (id === typeNameIdOfArrayOrSlice) {
+                let fsid = id;
+                if (fsid < 0) {
+                    fsid = null;
+                    if (!mgens) {
+                        return false;
+                    }
```
```suggestion
                    if (!mgens) {
                        return false;
                    }
                    fsid = null;
```

Changing the order a bit. :)

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -1698,10 +1749,24 @@ function initSearch(rawSearchIndex) {
 
             // If the result is too "bad", we return false and it ends this search.
-            if (!unifyFunctionTypes(row.type.inputs, parsedQuery.elems)) {
+            let mgens;
```
The `handleArgs` function became super huge. Would it be possible to move the callbacks you create in it outside as functions (with more args)? Or if you see another way to reduce its size, please do.

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -2314,8 +2441,15 @@ function initSearch(rawSearchIndex) {
             output = [];
         }
+        const where_clause = [];
+        const l = functionSearchType.length;
```
Please rename it (took me a few seconds when reading the line below than `i < l` was not `i < 1`, damn font, not the main reason though, but at least my brain will stopped to be confused by this haha).

GuillaumeGomez commented on tests/rustdoc-gui/search-corrections.goml:
```diff
@@ -54,3 +54,45 @@ assert-text: (
+})
+assert-text: (
+    ".search-corrections",
+    "Type \"notablestructwithlongnamr\" not found and used as generic parameter. Consider searching for \"notablestructwithlongname\" instead."
```
If you want you can also use single quotes so you won't need to escape double quotes.

GuillaumeGomez reviewed (changes requested):
Mostly code improvement suggestions. Nice work!

bors commented:
:pushpin: Commit 3c5ada25f748837363ceacda813d1d2d5474c79e has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez 3c5ada25f748837363ceacda813d1d2d5474c79e -->
<!-- homu: {"type":"Approved","sha":"3c5ada25f748837363ceacda813d1d2d5474c79e","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GuillaumeGomez commented:
Since https://github.com/rust-lang/rust/pull/112740 was merged, you'll need to update `search.js` to link to the correct location for the search docs.

bors commented:
:umbrella: The latest upstream changes (presumably #108537) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

notriddle committed:
rustdoc: bug fix for `-> option<t>`

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5772164701/job/15646852175) [(plain)](https://github.com/rust-lang/rust/commit/4c3e6fdbad74f62e0677b593695c3866026ebe36/checks/15646852175/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=notriddle
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_58a4ea70-8c7b-40a4-8bd7-1e794491e930
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=notriddle/advanced-search
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_58a4ea70-8c7b-40a4-8bd7-1e794491e930
GITHUB_REF=refs/pull/112725/merge
GITHUB_REF_NAME=112725/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=026bdcb0f0fa2bfd3f6d8a1967c910f143a4ee8d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_58a4ea70-8c7b-40a4-8bd7-1e794491e930
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_58a4ea70-8c7b-40a4-8bd7-1e794491e930
GITHUB_TRIGGERING_ACTOR=notriddle
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/112725/merge
GITHUB_WORKFLOW_SHA=026bdcb0f0fa2bfd3f6d8a1967c910f143a4ee8d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
---- [js-doc-test] tests/rustdoc-js/generics-trait.rs stdout ----

error: rustdoc-js test failed!
status: exit status: 1
command: "/usr/bin/node" "/checkout/src/tools/rustdoc-js/tester.js" "--doc-folder" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-js/generics-trait" "--crate-name" "generics_trait" "--test-file" "/checkout/tests/rustdoc-js/generics-trait.js"
Testing /checkout/tests/rustdoc-js/generics-trait.js ... FAILED
Testing /checkout/tests/rustdoc-js/generics-trait.js ... FAILED
[ query `Result<SomeTraiz>`]==> Expected exactly 0 results but found 2 in 'in_args'
[ query `Result<SomeTraiz>`]==> Expected exactly 0 results but found 2 in 'returned'
stderr: none



```

</details>

notriddle committed:
rustdoc: fix test case for generics that look like names

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5772414101/job/15647373729) [(plain)](https://github.com/rust-lang/rust/commit/9c19ed44739202fbab8ea5c9dc13d32c3739dea4/checks/15647373729/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=notriddle
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_88d95282-8687-4117-9f51-f61640faad58
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=notriddle/advanced-search
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_88d95282-8687-4117-9f51-f61640faad58
GITHUB_REF=refs/pull/112725/merge
GITHUB_REF_NAME=112725/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=1b3000edbffa117d15097eb3170c1b645360ab8c
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_88d95282-8687-4117-9f51-f61640faad58
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_88d95282-8687-4117-9f51-f61640faad58
GITHUB_TRIGGERING_ACTOR=notriddle
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/112725/merge
GITHUB_WORKFLOW_SHA=1b3000edbffa117d15097eb3170c1b645360ab8c
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
.......... (100/107)
......    (107/107)


/checkout/tests/rustdoc-gui/search-corrections.goml search-corrections... FAILED
[ERROR] (line 84) ".search-corrections" not found: for command `assert-css: (".search-corrections", {
    "display": "block"
})`
[ERROR] (line 87) ".search-corrections" not found: for command `assert-text: (
    ".search-corrections",
    "Type \"notablestructwithlongnamr\" not found and used as generic parameter. Consider searching for \"notablestructwithlongname\" instead."
)`
[ERROR] (line 92) ".error" not found: for command `assert-css: (".error", {
    "display": "block"
})`
[ERROR] (line 95) ".error" not found: for command `assert-text: (
    ".error",
    "Query parser error: \"Generic type parameter notablestructwithlongnamr does not accept generic parameters\"."

FError: ()
Build completed unsuccessfully in 0:08:03
##[group]Clock drift check
```

</details>

notriddle committed:
rustdoc: update tests for generic parsing and correction

bors commented:
:umbrella: The latest upstream changes (presumably #115436) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

notriddle committed:
Update docs since path-based type search works now

notriddle committed:
rustdoc-search: fix bugs when unboxing and reordering combine

notriddle committed:
rustdoc-doc: add `next_chunk` to list of `vec::intoiter<T> -> [T]`

This didn't show up before, because of some unification bugs
that were fixed in 269cb579479ab950e85a2e4078810501c29d7465

GuillaumeGomez commented:
Thanks, it's a very nice improvement!

@bors r+

bors commented:
:pushpin: Commit 4cf06e84abeb8bfc14fd7e9055db1c8ccf82266d has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez 4cf06e84abeb8bfc14fd7e9055db1c8ccf82266d -->
<!-- homu: {"type":"Approved","sha":"4cf06e84abeb8bfc14fd7e9055db1c8ccf82266d","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
